### PR TITLE
transform: performance like it matters

### PR DESCRIPTION
**Pull request by soypat**
Lot of noise in this PR, sorry about that. What matters is that
* I'm using a 256byte table for complement calculation instead of map
* I'm avoiding heap allocations where possible by using a byte slice and unsafe conversion.

```go
// ReverseComplement takes the reverse complement of a sequence.
// ReverseComplement expects ASCII input.
func ReverseComplement(sequence string) string {
    n := len(sequence)
    newSeq := make([]byte, n)
    for i := 0; i < n; i++ {
        newSeq[i] = complementTable[sequence[n-i-1]]
    }
    // This is how strings.Builder works with the String() method. If Mr. Go says it's safe...
    return *(*string)(unsafe.Pointer(&newSeq))
}
```

### CODE REVIEW

**Comment by soypat** [comment]
Test is failing because I included a test for the complement table and it seems it was doing something I did not expect. 'U's complement is  'A'?

File: transform/examples_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package transform_test
+
+import (
+	"fmt"
```
Comment (Koeng101) [inline transform/examples_test.go:4]: Splitting into a separate file for examples + tests is great

File: transform/transform.go [hunk 2]
```diff
@@ -12,74 +12,83 @@ ReverseComplement takes the reverse complement of a sequence.
...
-	for _, base := range complementString {
-		length--
-		newString[length] = base
+	n := len(sequence)
+	newSeq := make([]byte, n)
+	for i := 0; i < n; i++ {
+		newSeq[i] = complementTable[sequence[n-i-1]]
 	}
-	return string(newString)
+	// This is how strings.Builder works with the String() method. If Mr. Go says it's safe...
```
Comment (Koeng101) [inline transform/transform.go:25]: Useful comment

File: transform/transform.go [hunk 3]
```diff
@@ -12,74 +12,83 @@ ReverseComplement takes the reverse complement of a sequence.
...
 func ReverseComplement(sequence string) string {
-	complementString := strings.Map(ComplementBase, sequence)
-	length := len(complementString)
-	newString := make([]rune, length)
-	for _, base := range complementString {
-		length--
-		newString[length] = base
+	n := len(sequence)
+	newSeq := make([]byte, n)
+	for i := 0; i < n; i++ {
```
Comment (Koeng101) [inline transform/transform.go:22]: @TimothyStiles doesn't let any single letter variables in. Could you change the i and n to something longer / more descriptive? (like above)

File: transform/transform.go [hunk 4]
```diff
@@ -12,74 +12,83 @@ ReverseComplement takes the reverse complement of a sequence.
...
 // Reverse takes the reverse of a sequence.
 func Reverse(sequence string) string {
-	length := len(sequence)
-	newString := make([]rune, length)
-	for _, base := range sequence {
-		length--
-		newString[length] = base
+	n := len(sequence)
+	newSeq := make([]byte, n)
+	for i := 0; i < n; i++ {
```
Comment (Koeng101) [inline transform/transform.go:44]: Same as above. Single letter -> longer

File: transform/transform_test.go [hunk 5]
```diff
@@ -1,31 +1,97 @@
...
+}
 
-	// Output: ACATTAG
+func BenchmarkReverse(b *testing.B) {
+	rng := rand.New(rand.NewSource(1))
+	seq := randomSequence(rng)
+	for i := 0; i < b.N; i++ {
+		seq = Reverse(seq)
+	}
+	_ = seq
```
Comment (Koeng101) [inline transform/transform_test.go:96]: Is there a benchmark comparison to old method? Would like to see!

**Review by Koeng101 — APPROVED** [review]

**Comment by Koeng101** [comment]
A few things to change (single letter variables) - mostly interested in benchmarks. How much faster is it?

**Comment by TimothyStiles** [comment]
I can fix variables and stuff but I'm also interested in benchmarks! Would be happy to start on a real benchmark suite for Poly @soypat.

**Comment by soypat** [comment]
@Koeng101  @TimothyStiles I'll fix the variable names and add better documentation today- but I'd like to note that `i` as a generic counter in small scopes has been shown to be as readable or even more readable than a full length name. See [Beniamini et al.](https://www.cs.huji.ac.il/w~feit/papers/SingleLetter17ICPC.pdf).

It'd be awesome to have a benchmark suite like you say to get a real taste of how performance changes affects those using the library :smile:, more than happy to help out however I can. Been looking at refactoring the `fasta` package :eyes:.

<details><summary>Benchmarks for 32byte string input (click)</summary>

#### Before
```
goos: linux
goarch: amd64
pkg: github.com/TimothyStiles/poly/transform
cpu: Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
BenchmarkReverseComplement-8   	 1639549	       714.7 ns/op	     224 B/op	       3 allocs/op
BenchmarkComplement-8          	 3199111	       383.4 ns/op	      48 B/op	       1 allocs/op
BenchmarkReverse-8             	 4345659	       325.0 ns/op	     176 B/op	       2 allocs/op
PASS
```
#### After
```
BenchmarkReverseComplement-8   	18890924	        60.14 ns/op	      32 B/op	       1 allocs/op
BenchmarkComplement-8          	25396846	        55.45 ns/op	      32 B/op	       1 allocs/op
BenchmarkReverse-8             	19949305	        55.67 ns/op	      32 B/op	       1 allocs/op
```

</details>

File: transform/transform_test.go [hunk 6]
```diff
@@ -60,6 +60,13 @@ func TestReverseComplement(t *testing.T) {
 	}
 }
 
+func TestComplementBaseError(t *testing.T) {
+	complementBase := ComplementBase('!')
+	if complementBase != ' ' {
```
Comment (TimothyStiles) [inline transform/transform_test.go:65]: @Koeng101 if complement base can't find a complement for the given nucleotide what should it return? N?

File: transform/transform_test.go [hunk 7]
```diff
@@ -3,97 +3,86 @@ package transform
...
+	for _, testSequence := range evenAndOddLengthSequences {
+		got := ReverseComplement(testSequence)
+		expect := Reverse(Complement(testSequence))
 		if got != expect {
 			t.Errorf("mismatch with individual Reverse and Complement call:\n%q\n%q", got, expect)
 		}
 	}
 }
 
-func TestComplementBase(t *testing.T) {
```
Comment (soypat) [inline transform/transform_test.go]: Why was this test removed?

Comment (TimothyStiles) [reply transform/transform_test.go]: Added it back.

**Pull request merged**